pkg/operators/otel-metrics/otel-metrics.go

```diff
@@ -0,0 +1,485 @@
+			return float64(v)
+		}
+	case api.Kind_Float64:
+		return func(data datasource.Data) float64 {
+			v, _ := f.Float64(data)
+			return v
+		}
+	}
+}
+
+func (mc *metricsCollector) addKeyFunc(f datasource.FieldAccessor) error {
+	name := f.Name()
+	switch f.Type() {
+	default:
+		return fmt.Errorf("unsupported field type for metrics collector: %s", f.Type())
+	case api.Kind_String, api.Kind_CString:
+		mc.keys = append(mc.keys, func(data datasource.Data) attribute.KeyValue {
+			val, _ := f.String(data)
+			return attribute.KeyValue{Key: attribute.Key(name), Value: attribute.StringValue(val)}
+		})
+	case api.Kind_Uint8,
+		api.Kind_Uint16,
+		api.Kind_Uint32,
+		api.Kind_Uint64,
+		api.Kind_Int8,
+		api.Kind_Int16,
+		api.Kind_Int32,
+		api.Kind_Int64:
+		asIntFn := asInt64(f)
+		mc.keys = append(mc.keys, func(data datasource.Data) attribute.KeyValue {
+			return attribute.KeyValue{Key: attribute.Key(name), Value: attribute.Int64Value(asIntFn(data))}
+		})
+	case api.Kind_Float32, api.Kind_Float64:
+		asFloatFn := asFloat64(f)
+		mc.keys = append(mc.keys, func(data datasource.Data) attribute.KeyValue {
+			return attribute.KeyValue{Key: attribute.Key(name), Value: attribute.Float64Value(asFloatFn(data))}
+		})
+	}
+	return nil
+}
+
+func (mc *metricsCollector) addValFunc(f datasource.FieldAccessor, metricType string) error {
+	var options []metric.InstrumentOption
+	if description := f.Annotations()["metrics.description"]; description != "" {
+		options = append(options, metric.WithDescription(description))
+	}
+	if unit := f.Annotations()["metrics.unit"]; unit != "" {
+		options = append(options, metric.WithUnit(unit))
+	}
+
+	switch f.Type() {
+	default:
+		return fmt.Errorf("unsupported field type for metrics value %q: %s", f.Name(), f.Type())
+	case api.Kind_Uint8,
+		api.Kind_Uint16,
+		api.Kind_Uint32,
+		api.Kind_Uint64,
+		api.Kind_Int8,
+		api.Kind_Int16,
+		api.Kind_Int32,
+		api.Kind_Int64:
+		asIntFn := asInt64(f)
+		switch metricType {
+		case MetricTypeCounter:
+			tOptions := make([]metric.Int64CounterOption, 0)
+			for _, option := range options {
+				tOptions = append(tOptions, option)
+			}
+			ctr, err := mc.meter.Int64Counter(f.Name(), tOptions...)
+			if err != nil {
+				return fmt.Errorf("adding metric %s for %q: %w", metricType, f.Name(), err)
+			}
+			mc.values = append(mc.values, func(ctx context.Context, data datasource.Data, set attribute.Set) {
+				ctr.Add(ctx, asIntFn(data), metric.WithAttributeSet(set))
+			})
+		case MetricTypeGauge:
+			tOptions := make([]metric.Int64GaugeOption, 0)
+			for _, option := range options {
+				tOptions = append(tOptions, option)
+			}
+			ctr, err := mc.meter.Int64Gauge(f.Name(), tOptions...)
+			if err != nil {
+				return fmt.Errorf("adding metric %s for %q: %w", metricType, f.Name(), err)
+			}
+			mc.values = append(mc.values, func(ctx context.Context, data datasource.Data, set attribute.Set) {
+				ctr.Record(ctx, asIntFn(data), metric.WithAttributeSet(set))
+			})
+		}
```

Comment:
That would only solve it for `tOptions` (but I think you would also need to add some constraints as to what could be casted to T and what not - I haven't found a proper interface to use; and if using `T any`, I guess it would make it check only on runtime (and panic if incompatible), not compile time, right? (need to check on that)) - but there are also differences in `mc.meter.Int64Counter()`/`mc.meter.Int64Gauge()` and inside the callback it's `.Add()` vs `.Record()` with different function signatures, so I don't see this being solved easily, at least not in a way that makes the code more readable. Feel free to prove me wrong 😄

Reply:
Right, I've also tried it similarly to what you did with the funcPointer - but that `something` in your example has to have either `.Record()` or `.Add()` as method (not a shared method, sadly), depending on the type - the `metrics.WithAttributeSet()` has to be of type `AddOption` or `RecordOption`, depending on type of the metric as well. The biggest issue is that you have multiple generic types (at least one for the option type and another one for the metric type - probably yet another one for the last parameter to `.Record()`/`.Add()`. And then you'd have to correlate them, which is another issue.

I can imagine there being a solution if we were to implement our own types as aliases of Int64Gauge/Int64Counter with conversion methods implemented in them and maybe a combined method that will divert to either `.Add()` or `.Record()`, but that's far more complicated than those couple of similar lines, IMHO.

If there's an easy solution I don't see, I'm happy to learn about it - maybe I'm just not seeing the forest for the trees, hehe.